### PR TITLE
fix: NTRN-278. Fix contract tests after neutron migration to IS

### PR DESCRIPTION
**Pull request by albertandrejev**
Make contracts tests to work with latest version of neutron with CCV modules.

Depends on: https://github.com/neutron-org/neutron/pull/107

This PR was transferred to https://github.com/neutron-org/neutron-dev-contracts/pull/1, feel free to remove it after another PR will be merged.

### CODE REVIEW

File: scripts/test_crud.sh [hunk 1]
```diff
@@ -5,36 +5,49 @@ NEUTRON_DIR=${NEUTRON_DIR:-../neutron}
...
 KEY=demowallet1
+FAUCET=demowallet3
 ADMIN=neutron1m9l358xunhhwds0568za49mzhvuxx9ux8xafx2
-RLY2=neutron17dtl0mjt3t77kpuhg2edqzjpszulwhgzcdvagh
+RLY2=cosmos17dtl0mjt3t77kpuhg2edqzjpszulwhgzuj9ljs
+
 BIN=neutrond
+GAIAD_BIN=gaiad
+
+TEST_WALLET=test_wallet
```
Comment (sotnikov-s) [inline scripts/test_crud.sh:15]: didn't get this move. it was `KEY=demowallet1` from whose behalf all txs were executed, but you introduced `TEST_WALLET=test_wallet` (which also required initialisation) and used it as main actor instead. 1) what's the reason? 2) KEY became unused

File: scripts/test_custom_querier.sh [hunk 2]
```diff
@@ -4,11 +4,12 @@ CHAINID2=test-2
 NEUTRON_DIR=${NEUTRON_DIR:-../neutron}
 HOME=${NEUTRON_DIR}/data/test-1/
 HOME2=${NEUTRON_DIR}/data/test-2/
-KEY=demowallet1
+KEY=demowallet3
```
Comment (sotnikov-s) [inline scripts/test_custom_querier.sh:7]: in the `test_crud` and `ibc_transfer` scripts `demowallet3` is the faucet's account. why change it to the main actor here? this is confusing compared to the scenario where we have same roles for wallets across all tests

File: scripts/test_ibc_transfer.sh [hunk 3]
```diff
@@ -2,30 +2,39 @@ CONTRACT=./artifacts/ibc_transfer.wasm
...
 HOME=${NEUTRON_DIR}/data/test-1/
-HOME2=${NEUTRON_DIR}/data/test-2/
 KEY=demowallet1
 ADMIN=neutron1m9l358xunhhwds0568za49mzhvuxx9ux8xafx2
+
 BIN=neutrond
+GAIAD_BIN=gaiad
+
+FAUCET=demowallet3
+TEST_WALLET=test_wallet
```
Comment (sotnikov-s) [inline scripts/test_ibc_transfer.sh:12]: same about `KEY` substitution with `TEST_WALLET`

File: scripts/test_interchain_txs.sh [hunk 4]
```diff
@@ -9,41 +16,55 @@ NEUTRON_DIR=${NEUTRON_DIR:-../neutron}
...
+RES=$(${BIN} tx wasm execute $CONTRACT_ADDRESS "{\"register\": {\"connection_id\": \"connection-0\", \"interchain_account_id\": \"test\"}}" --from ${TEST_ADDR}  -y --chain-id ${CHAIN_ID_1} --output json --broadcast-mode=block --gas-prices 0.0025stake --gas 1000000 --keyring-backend test --home ${HOME_1} --node tcp://127.0.0.1:16657)
 echo $RES
-sleep 10
 
-RES=$(curl http://127.0.0.1:1316/wasm/contract/$CONTRACT_ADDRESS/smart/eyJpbnRlcmNoYWluX2FjY291bnRfYWRkcmVzc19mcm9tX2NvbnRyYWN0Ijp7ImludGVyY2hhaW5fYWNjb3VudF9pZCI6InRlc3QifX0\=?encoding\=base64 | jq -r ".result.smart")
+wait_for_tx "http://127.0.0.1:1316/wasm/contract/$CONTRACT_ADDRESS/smart/eyJpbnRlcmNoYWluX2FjY291bnRfYWRkcmVzc19mcm9tX2NvbnRyYWN0Ijp7ImludGVyY2hhaW5fYWNjb3VudF9pZCI6InRlc3QifX0\=?encoding\=base64" ".result.smart"
+RES=$FUNC_RETURN
+
+#RES=$(curl http://127.0.0.1:1316/wasm/contract/$CONTRACT_ADDRESS/smart/eyJpbnRlcmNoYWluX2FjY291bnRfYWRkcmVzc19mcm9tX2NvbnRyYWN0Ijp7ImludGVyY2hhaW5fYWNjb3VudF9pZCI6InRlc3QifX0\=?encoding\=base64 | jq -r ".result.smart")
```
Comment (sotnikov-s) [inline scripts/test_interchain_txs.sh:53]: can be removed?

File: scripts/test_interchain_txs.sh [hunk 5]
```diff
@@ -9,41 +16,55 @@ NEUTRON_DIR=${NEUTRON_DIR:-../neutron}
...
+
 #Register interchain account
-RES=$(${BIN} tx wasm execute $CONTRACT_ADDRESS "{\"register\": {\"connection_id\": \"connection-0\", \"interchain_account_id\": \"test\"}}" --from ${ADDRESS_1}  -y --chain-id ${CHAIN_ID_1} --output json --broadcast-mode=block --gas-prices 0.0025stake --gas 1000000 --keyring-backend test --home ${HOME_1} --node tcp://127.0.0.1:16657)
+RES=$(${BIN} tx wasm execute $CONTRACT_ADDRESS "{\"register\": {\"connection_id\": \"connection-0\", \"interchain_account_id\": \"test\"}}" --from ${TEST_ADDR}  -y --chain-id ${CHAIN_ID_1} --output json --broadcast-mode=block --gas-prices 0.0025stake --gas 1000000 --keyring-backend test --home ${HOME_1} --node tcp://127.0.0.1:16657)
 echo $RES
-sleep 10
 
-RES=$(curl http://127.0.0.1:1316/wasm/contract/$CONTRACT_ADDRESS/smart/eyJpbnRlcmNoYWluX2FjY291bnRfYWRkcmVzc19mcm9tX2NvbnRyYWN0Ijp7ImludGVyY2hhaW5fYWNjb3VudF9pZCI6InRlc3QifX0\=?encoding\=base64 | jq -r ".result.smart")
+wait_for_tx "http://127.0.0.1:1316/wasm/contract/$CONTRACT_ADDRESS/smart/eyJpbnRlcmNoYWluX2FjY291bnRfYWRkcmVzc19mcm9tX2NvbnRyYWN0Ijp7ImludGVyY2hhaW5fYWNjb3VudF9pZCI6InRlc3QifX0\=?encoding\=base64" ".result.smart"
```
Comment (sotnikov-s) [inline scripts/test_interchain_txs.sh:50]: isn't it possible to get the same result not with `curl` but with `neutrond query wasm contract-state`?

File: scripts/test_tx_query_result.sh [hunk 6]
```diff
@@ -10,37 +11,44 @@ HOME_1=${NEUTRON_DIR}/data/test-1/
...
 USERNAME_2=demowallet2
-KEY_2=$(neutrond keys show demowallet2 -a --keyring-backend test --home ${HOME_2})
+KEY_2=$(gaiad keys show demowallet2 -a --keyring-backend test --home ${HOME_2})
 ADMIN=$(neutrond keys show demowallet1 -a --keyring-backend test --home ${HOME_1})
 
-TARGET_ADDRESS=neutron1mjk79fjjgpplak5wq838w0yd982gzkyf8fxu8u
-VAL2=neutronvaloper1qnk2n4nlkpw9xfqntladh74w6ujtulwnqshepx
+TARGET_ADDRESS=cosmos17dtl0mjt3t77kpuhg2edqzjpszulwhgzuj9ljs
+VAL2=cosmosvaloper1qnk2n4nlkpw9xfqntladh74w6ujtulwnmxnh3k
+TEST_WALLET=test_wallet
```
Comment (sotnikov-s) [inline scripts/test_tx_query_result.sh:21]: same about `USERNAME_1` substitution with `TEST_WALLET`

File: scripts/test_crud.sh (hunk 1 above)
Comment (albertandrejev) [reply scripts/test_crud.sh:15]: demowallet1 is used to run query relayer as well, in this case we will have two process that are using same address to produce transaction and it leads us to a lot of `sequence` problems.

File: scripts/test_custom_querier.sh (hunk 2 above)
Comment (albertandrejev) [reply scripts/test_custom_querier.sh:7]: You are right I need to modify code to use faucet address everywhere.

File: scripts/test_crud.sh (hunk 1 above)
Comment (albertandrejev) [reply scripts/test_crud.sh:15]: Also dedicated address per script run helps us to have more or less clean environment between script executions.

File: scripts/test_interchain_txs.sh (hunk 5 above)
Comment (albertandrejev) [reply scripts/test_interchain_txs.sh:50]: Yes, possible. But it was here before, so I decided not to change used approach. Do you think it is better to change it to `neutrond ...` ?

**Review by oopcode — DISMISSED** [review]

File: scripts/test_interchain_txs.sh (hunk 5 above)
Comment (sotnikov-s) [reply scripts/test_interchain_txs.sh:50]: yep, `curl` is a workaround whereas `neutrond` is straightforward. also, we use direct `neutrond` calls everywhere

File: scripts/test_crud.sh (hunk 1 above)
Comment (sotnikov-s) [reply scripts/test_crud.sh:15]: makes sense, thanks

Comment (foxpy) [reply scripts/test_crud.sh:15]: > demowallet1 is used to run query relayer as well, in this case we will have two process that are using same address to produce transaction and it leads us to a lot of `sequence` problems.

maybe we should finally allocate an account for relayer? I am sick tired of this collision. I have proposed to do that N months ago but we decided to postpone it.

Comment (foxpy) [reply scripts/test_crud.sh:15]: and I will do it if we approve adding one more account :^)

**Review by foxpy — CHANGES_REQUESTED** [review]
While we are here, please add
```bash
# http://redsymbol.net/articles/unofficial-bash-strict-mode/
set -euo pipefail
IFS=$'\n\t'
```
in the beginning of every script. It will help us seeing errors early without having to scratch our heads reading obscure walls of errors.

You will probably run into various kinds of issues while applying this patch, so feel free to ask me for help! You can also consult http://redsymbol.net/articles/unofficial-bash-strict-mode/, it provides really nice instructions.

File: scripts/test_crud.sh (hunk 1 above)
Comment (albertandrejev) [reply scripts/test_crud.sh:15]: @foxpy I think we can achieve this in other PR, I will prepare task for Jira in the meantime

Comment (albertandrejev) [reply scripts/test_crud.sh:15]: Added https://p2pvalidator.atlassian.net/browse/NTRN-297

**Comment by albertandrejev** [comment]
> While we are here, please add
> 
> ```shell
> # http://redsymbol.net/articles/unofficial-bash-strict-mode/
> set -euo pipefail
> IFS=$'\n\t'
> ```

Also this one can go to another task

**Review by foxpy — APPROVED** [review]
utACK

**Review by foxpy — CHANGES_REQUESTED** [review]
`test_ibc_transfer.sh` returns this on my side:
```
Failures before test. Should be empty
{
  "failures": [],
  "pagination": {
    "next_key": null,
    "total": "0"
  }
}
Show failures after test
{
  "failures": [],
  "pagination": {
    "next_key": null,
    "total": "0"
  }
}
```

doesn't look right.

**Review by foxpy — APPROVED** [review]

**Review by pr0n00gler — APPROVED** [review]

**Pull request closed without merging**